### PR TITLE
fix: pre-scan markdown titles for group-icons to fix missing icons in production

### LINKED ISSUE #600: 【BUG】代码块标题使用文件拓展名语法不显示图标
### Describe the bug

如题，这个bug现在又来了，将框架和主题版本升级了之后就又出现了

### Reproduction

依旧如此

### System Info

```Shell
win
```

### Used Package Manager

pnpm

### Validations

- [x] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A, please open a GitHub Discussion instead.
- [x] The provided reproduction is a [minimal reproducible](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by YunYouJun**
## Summary

- Pre-scans all markdown pages at startup to extract code block titles (`[title]` syntax in fenced code blocks)
- Passes extracted titles as `defaultLabels` to `groupIconVitePlugin`, ensuring icon CSS is always generated
- Adds unit tests for the title extraction utility

## Root Cause

During SSG builds, the `vitepress-plugin-group-icons` Vite plugin's `transform` hook may not fire for all markdown files because file IDs don't always match its filter pattern (`/\.(md|md\?vue|md\?v=)$/`) after Vue SFC compilation by `@vitejs/plugin-vue`. This causes the `matches` Set to remain empty, so `generateCSS()` produces no icon-specific CSS — resulting in missing code block title icons in production while they work fine in dev mode.

## Fix

Instead of relying on the `transform` hook to collect `data-title` values at runtime, we pre-scan all markdown source files during plugin initialization and extract code block titles using a regex. These titles are passed as `defaultLabels` to `groupIconVitePlugin`, which merges them into the CSS generation unconditionally.

This approach is robust because:
- It doesn't depend on Vite's transform pipeline ordering or file ID patterns
- The scanning runs concurrently with other async plugin initializations (no added latency)
- It works identically in both dev and build modes

## Performance

- 93 markdown files scanned in **~5ms** (benchmarked on demo site)
- Runs in `Promise.all` alongside shiki highlighter init (~500ms+), so **zero observable startup latency**
- Even 1000+ posts would only take ~50ms, still hidden behind the shiki bottleneck
- Regex uses `[^\S\n]+` instead of `\s+` to prevent cross-line false matches (e.g. `[[toc]]` after a code fence)

## Test plan

- [x] Unit tests for `extractCodeBlockTitles` utility (11 tests covering various markdown patterns including edge cases)
- [x] All existing tests pass (112/112)
- [x] ESLint passes cleanly
- [x] Build succeeds

Closes #600

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/valaxy/node/utils/groupIcons.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * Regex to extract code block titles from markdown fenced code blocks.
+ *
+ * Matches patterns like:
+ * - ` ```ts [filename.ts] ` → captures "filename.ts"
+ * - ` ```typescript [builtin.ts] ` → captures "builtin.ts"
+ * - ` ```sh [pnpm] ` → captures "pnpm"
+ *
+ * Also handles nested brackets like `[foo[bar]]`.
+ */
+const codeBlockTitleRE = /^`{3}\S*\s+\[((?:[^[\]]|\[[^[\]]*\])*)\]/gm
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts]: Fixed: Changed regex to `{3}` with `[^\n[]*` — the character class already consumes extra backticks beyond the initial 3, so `{3,}` is unnecessary. The linter also flagged the original `{3,}` + `[^\n[]*` combination as having potential super-linear backtracking. Quadruple backtick test still passes.

File: packages/valaxy/node/utils/groupIcons.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  const titles = new Set<string>()
+
+  await Promise.all(
+    options.pages.map(async (page) => {
+      try {
+        const content = await readFile(resolve(pagesDir, page), 'utf-8')
+        extractCodeBlockTitles(content, titles)
+      }
+      catch {
+        // Ignore files that can't be read
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts]: Fixed: Replaced silent catch with `debug()` logging via the `debug` package (consistent with other Valaxy modules like `valaxy:md`). Errors are now visible when running with `DEBUG=valaxy:group-icons`.

File: packages/valaxy/node/utils/groupIcons.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    }),
+  )
+
+  return [...titles]
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts]: Fixed: Now returns `[...titles].sort()` for deterministic output.

File: test/group-icons.test.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      '',
+      '```vue [App.vue]',
+      '<template></template>',
+      '```',
+    ].join('\n')
+    const titles = extractCodeBlockTitles(md)
+    expect([...titles]).toContain('app.ts')
+    expect([...titles]).toContain('App.vue')
+  })
+
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply test/group-icons.test.ts:25]: Fixed: Added two tests — `handles code blocks with meta attrs before the title` and `handles code blocks with meta attrs after the title`. Also updated the regex from `\S*[^\S\n]+` to `[^\n[]*` so it correctly matches titles preceded by attrs like `{7-9}`.

File: packages/valaxy/node/utils/groupIcons.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    options.pages.map(async (page) => {
+      try {
+        const content = await readFile(resolve(pagesDir, page), 'utf-8')
+        extractCodeBlockTitles(content, titles)
+      }
+      catch {
+        // Ignore files that can't be read
+      }
+    }),
+  )
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts]: Fixed: Replaced `Promise.all` with `pMap` (from `p-map`, already a dependency) with `concurrency: 64`, consistent with how `localSearchPlugin` handles page scanning.

**Comment by YunYouJun** [comment]
## Copilot Review Loop — Summary

PR: #648 (https://github.com/YunYouJun/valaxy/pull/648)
Cycles completed: 1

| # | File:Line | Issue | Category | Action | Reply | Cycle |
|---|-----------|-------|----------|--------|-------|-------|
| 1 | `groupIcons.ts:15` | `{3}` vs `{3,}` for quadruple backticks | SHOULD_FIX | Fixed — use `{3}` with `[^\n[]*` (linter-optimal, equivalent behavior) | ✅ | 1 |
| 2 | `groupIcons.ts:39` | Silent catch hides unreadable files | SHOULD_FIX | Fixed — added `debug()` logging via `debug` package | ✅ | 1 |
| 3 | `groupIcons.ts:44` | Non-deterministic Set ordering | SHOULD_FIX | Fixed — `[...titles].sort()` | ✅ | 1 |
| 4 | `group-icons.test.ts:25` | Missing test for meta attrs + title | SHOULD_FIX | Fixed — added 2 tests for `{7-9}` before/after title | ✅ | 1 |
| 5 | `groupIcons.ts:42` | Unbounded `Promise.all` EMFILE risk | SHOULD_FIX | Fixed — replaced with `pMap` (concurrency: 64) | ✅ | 1 |

Commits:
- `958e3a4e` fix: pre-scan markdown titles for group-icons to fix missing icons in production
- `f2007204` fix: address copilot review feedback

Remaining items needing human review: None

File: packages/valaxy/node/utils/groupIcons.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * Regex to extract code block titles from markdown fenced code blocks.
+ *
+ * Matches patterns like:
+ * - ` ```ts [filename.ts] ` → captures "filename.ts"
+ * - ` ```typescript [builtin.ts] ` → captures "builtin.ts"
+ * - ` ```sh [pnpm] ` → captures "pnpm"
+ * - ` ```ts {7-9} [site.config.ts] ` → captures "site.config.ts"
+ *
+ * Uses `[^\S\n]+` instead of `\s+` to prevent cross-line false matches
+ * (e.g. `[[toc]]` on the line after a code fence).
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts]: Fixed ✅ — Updated the doc comment to accurately describe the current regex pattern `[^\n[]*` instead of the outdated `[^\S\n]+`.

File: packages/valaxy/node/utils/groupIcons.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+ * Matches patterns like:
+ * - ` ```ts [filename.ts] ` → captures "filename.ts"
+ * - ` ```typescript [builtin.ts] ` → captures "builtin.ts"
+ * - ` ```sh [pnpm] ` → captures "pnpm"
+ * - ` ```ts {7-9} [site.config.ts] ` → captures "site.config.ts"
+ *
+ * Uses `[^\n[]*` to match any content on the same line before `[title]`,
+ * preventing cross-line false matches (e.g. `[[toc]]` after a code fence).
+ */
+const codeBlockTitleRE = /^`{3}[^\n[]*\[((?:[^[\]]|\[[^[\]]*\])*)\]/gm
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts]: Fixed ✅ — Added `[ \t]{0,3}` before the backticks to allow up to 3 leading spaces/tabs per CommonMark spec. Also added tests for both the 3-space case (matches) and the 4-space case (does not match).

File: packages/valaxy/node/utils/groupIcons.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  const pagesDir = resolve(options.userRoot, 'pages')
+  const titles = new Set<string>()
+
+  await pMap(
+    options.pages,
+    async (page) => {
+      try {
+        const content = await readFile(resolve(pagesDir, page), 'utf-8')
+        extractCodeBlockTitles(content, titles)
+      }
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts:45]: Skipped — The `@include` directive is a niche feature and expanding includes here would require importing and running the full include transform pipeline, adding significant complexity. The plugin `transform` hook still handles most cases in dev, and this pre-scan is a production safety net. Users with titles in included files can add them to `groupIcons.defaultLabels` in their config.

File: test/group-icons.test.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
+import { describe, expect, it } from 'vitest'
+import { extractCodeBlockTitles } from '../packages/valaxy/node/utils/groupIcons'
+
+describe('extractCodeBlockTitles', () => {
+  it('extracts title from fenced code block', () => {
+    const md = '```ts [filename.ts]\nconsole.log("hello")\n```'
+    const titles = extractCodeBlockTitles(md)
+    expect([...titles]).toEqual(['filename.ts'])
+  })
+
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply test/group-icons.test.ts:10]: Fixed ✅ — Added two tests: one for a 3-space indented fence (should match) and one for a 4-space indented fence (should not match, as CommonMark treats it as a code block).

File: packages/valaxy/node/utils/groupIcons.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      catch (error) {
+        debug('Failed to read %s: %O', page, error)
+      }
+    },
+    { concurrency: 64 },
```
[the remark this answers is not in the document]
Comment (YunYouJun) [reply packages/valaxy/node/utils/groupIcons.ts:50]: Skipped — The concurrency difference is intentional. `localSearchPlugin` uses 10 because each iteration does heavy markdown rendering (`render(file)` → full HTML). Here we only do `readFile` + a trivial regex match, which is orders of magnitude lighter. 64 concurrent file reads is well within typical OS FD limits (default 1024+) and keeps the scan fast. Adding a comment explaining this would be over-documenting an implementation detail.
